polkadot/xcm/xcm-builder/src/tests/barriers.rs

```diff
@@ -532,3 +532,239 @@ fn allow_hrmp_notifications_from_relay_chain_should_work() {
+	/// contains a single `UnsubscribeVersion`
+	struct DenyUnsubscribeVersionAsStackLimitReached;
+	impl DenyExecution for DenyUnsubscribeVersionAsStackLimitReached {
+		fn deny_execution<RuntimeCall>(
+			_origin: &Location,
+			instructions: &mut [Instruction<RuntimeCall>],
+			_max_weight: Weight,
+			_properties: &mut Properties,
+		) -> Option<ProcessMessageError> {
+			if instructions.len() != 1 {
+				return None;
+			}
+			match instructions.get(0).unwrap() {
+				UnsubscribeVersion { .. } => Some(ProcessMessageError::StackLimitReached),
+				_ => None,
+			}
+		}
+	}
+
+	/// A dummy `ShouldExecute` impl which returns `Ok(())` when XCM contains a single `ClearError`,
+	/// else return `ProcessMessageError::Yield`
+	struct AllowSingleClearErrorOrYield;
+	impl ShouldExecute for AllowSingleClearErrorOrYield {
+		fn should_execute<Call>(
+			_origin: &Location,
+			instructions: &mut [Instruction<Call>],
+			_max_weight: Weight,
+			_properties: &mut Properties,
+		) -> Result<(), ProcessMessageError> {
+			instructions.matcher().assert_remaining_insts(1)?.match_next_inst(
+				|inst| match inst {
+					ClearError { .. } => Ok(()),
+					_ => Err(ProcessMessageError::Yield),
+				},
+			)?;
+			Ok(())
+		}
+	}
+
+	/// A dummy `ShouldExecute` impl which returns `Ok(())` when XCM contains `ClearTopic` and
+	/// origin from `Here`, else return `ProcessMessageError::Unsupported`
+	struct AllowClearTopicFromHere;
+	impl ShouldExecute for AllowClearTopicFromHere {
+		fn should_execute<Call>(
+			origin: &Location,
+			instructions: &mut [Instruction<Call>],
+			_max_weight: Weight,
+			_properties: &mut Properties,
+		) -> Result<(), ProcessMessageError> {
+			ensure!(origin.clone() == Here.into_location(), ProcessMessageError::Unsupported);
+			let mut found = false;
+			instructions.matcher().match_next_inst_while(
+				|_| true,
+				|inst| match inst {
+					ClearTopic { .. } => {
+						found = true;
+						Ok(ControlFlow::Break(()))
+					},
+					_ => Ok(ControlFlow::Continue(())),
+				},
+			)?;
+			ensure!(found, ProcessMessageError::Unsupported);
+			Ok(())
+		}
+	}
+	// closure for (xcm, origin) testing with `DenyThenTry`
+	let assert_should_execute = |mut xcm: Vec<Instruction<()>>, origin, expected_result| {
+		pub type Barrier = DenyThenTry<
+			(
+				DenyClearTransactStatusAsYield,
+				DenyClearOriginFromHereAsBadFormat,
+				DenyUnsubscribeVersionAsStackLimitReached,
+			),
+			(AllowSingleClearErrorOrYield, AllowClearTopicFromHere),
+		>;
+		assert_eq!(
+			Barrier::should_execute(
+				&origin,
+				&mut xcm,
+				Weight::from_parts(10, 10),
+				&mut props(Weight::zero()),
+			),
+			expected_result
+		);
+	};
+
+	// Deny cases:
+	// trigger DenyClearTransactStatusAsYield
+	assert_should_execute(
+		vec![ClearTransactStatus],
+		Parachain(1).into_location(),
+		Err(ProcessMessageError::Yield),
+	);
+	// DenyClearTransactStatusAsYield wins against AllowClearErrorOrYield
```

Comment:
```suggestion
	// DenyClearTransactStatusAsYield wins against AllowSingleClearErrorOrYield
```

Reply:
As you have two instructions, shouldn't this fail anyway just with `AllowSingleClearErrorOrYield`?